src/System.Xml.XDocument/src/System/Xml/Linq/Extensions.cs

```diff
@@ -269,7 +268,19 @@ public static IEnumerable<XElement> DescendantsAndSelf(this IEnumerable<XElement
         /// </returns>
         public static IEnumerable<T> InDocumentOrder<T>(this IEnumerable<T> source) where T : XNode
         {
-            return Enumerable.OrderBy(source, n => (XNode)n, XNode.DocumentOrderComparer);
+            if (source == null) throw new ArgumentNullException("source");
+            return DocumentOrderIterator<T>(source);
+        }
+        
+        private static IEnumerable<T> DocumentOrderIterator<T>(IEnumerable<T> source) where T : XNode
+        {
+            int count;
+            T[] items = EnumerableHelpers.ToArray(source, out count);
+            if (count > 0)
+            {
+                Array.Sort(items, 0, count, XNode.DocumentOrderComparer);
+                for (int i = 0; i != count; ++i) yield return items[i];
```

Comment:
Though thinking about it further, without more work that would move the ToArray work to be done during the initial call rather than during the first MoveNext.  Ok, let's leave it as is.